pgx/src/trigger_support/pg_trigger.rs

```diff
@@ -0,0 +1,210 @@
+use crate::{
+    heap_tuple::PgHeapTuple,
+    pg_sys,
+    pgbox::{AllocatedByPostgres, PgBox},
+    rel::PgRelation,
+    trigger_support::{
+        called_as_trigger, PgTriggerError, PgTriggerLevel, PgTriggerOperation, PgTriggerSafe,
+        PgTriggerWhen, TriggerEvent, TriggerTuple,
+    },
+};
+use cstr_core::c_char;
+use std::borrow::Borrow;
+
+/**
+The datatype accepted by a trigger
+
+A safe structure providing the an API similar to the constants provided in a PL/pgSQL function.
+
+Usage examples exist in the module level docs.
+*/
+pub struct PgTrigger {
+    trigger_data: PgBox<pgx_pg_sys::TriggerData>,
+    #[allow(dead_code)]
+    fcinfo: pg_sys::FunctionCallInfo,
+}
+
+impl PgTrigger {
+    pub unsafe fn from_fcinfo(fcinfo: pg_sys::FunctionCallInfo) -> Result<Self, PgTriggerError> {
+        if fcinfo.is_null() {
+            return Err(PgTriggerError::NullFunctionCallInfo);
+        }
+        if !called_as_trigger(fcinfo) {
+            return Err(PgTriggerError::NotTrigger);
+        }
+        let fcinfo_data = &*fcinfo;
+
+        if fcinfo_data.context.is_null() {
+            return Err(PgTriggerError::NullTriggerData);
+        }
+        let trigger_data: PgBox<pg_sys::TriggerData> =
+            PgBox::from_pg(fcinfo_data.context as *mut pg_sys::TriggerData);
+
+        Ok(Self {
+            trigger_data,
+            fcinfo,
+        })
+    }
+
+    /// A reference to the underlaying trigger data
+    pub fn trigger_data(&self) -> &pgx_pg_sys::TriggerData {
+        self.trigger_data.borrow()
+    }
+
+    /// A reference to the underlaying fcinfo
+    pub fn fcinfo(&self) -> &pg_sys::FunctionCallInfo {
+        self.fcinfo.borrow()
+    }
+
+    /// The new HeapTuple
+    // Derived from `pgx_pg_sys::TriggerData.tg_newtuple` and `pgx_pg_sys::TriggerData.tg_newslot.tts_tupleDescriptor`
+    pub unsafe fn new(&self) -> Option<PgHeapTuple<'_, AllocatedByPostgres>> {
```

Comment:
I really feel like these functions don't need to be `unsafe`.

The `::from_fcinfo()` function definitely does, but at that point the developer is saying "hey compiler, I decided what I gave you is correct".  So for us to continue to not trust that fcinfo object seems wrong to me.

Think about things like `String::from_utf8_unchecked()`.  Yes, that function is unsafe, but `String` itself is safe and otherwise trusts the backing Vec.

Reply:
Do you also think `PgHeapTuple::from_trigger_data` should be safe?

Reply:
I think if we check `trigger_data.tg_trigger` during `from_fcinfo` we can make `name()`, `old_transition_table_name()`, and `new_transition_table_name()` safe.

Reply:
No, that should absolutely be unsafe because it can't prove the pointer is correct.  But once the user calls it as:

```rust
let pg_trigger = unsafe { PgTrigger::from_trigger_data(...) };
```

I think we're clear to assume whatever was given to us is valid -- after all, the programmer told us so b/c the used `unsafe {}`.

Reply:
`PgRelation::open` is `unsafe` itself, I can't really see us making that safe unless we eagerly open it during `from_fcinfo`.

Reply:
Sorry, how is `PgRelation` involved here?

Reply:
`table_name` and `table_schema` are got via `PgRelation::open` on `fcinfo.trigger_data.relation_data.rd_id` which is unsafe.

Reply:
I'm confused.  I'm saying that `PgTrigger::from_trigger_data(...)` **should** be `unsafe`.  As such, the programmer is telling us the provided pointer is properly defined.  How does that impact that function's usage of `PgReation::open()`?

Reply:
Well, to start the `from_trigger_data` function you're referring to has never existed.. 😅 

I sat down and wrote some safety rules on the `from_fcinfo` function and I'm now making several more functions safe because the user has... well, agreed to my terms.